### PR TITLE
Print the transform name if exception has occurred in the ParDo evaluator

### PR DESCRIPTION
**Pull request by mairbek**
R:@jkff, please take a look

### CODE REVIEW

**Comment by mairbek** [comment]
Added a test.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java [hunk 1]
```diff
@@ -495,8 +495,36 @@ public void unreachable(ProcessContext c) {
...
+  public void testUserException() throws Exception {
+    Pipeline p = getPipeline();
+    PCollection<Long> pcollection = p.apply("Step 1", Create.of(1))
+        .apply("Step 2", ParDo.of(new DoFn<Integer, Long>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            throw new IllegalStateException("Custom exception element " + c.element());
+          }
+        })).setCoder(VarLongCoder.of());
+    pcollection
```
Comment (jkff) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java]: This seems unrelated to the current test

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java [hunk 2]
```diff
@@ -495,8 +495,36 @@ public void unreachable(ProcessContext c) {
...
+                  @ProcessElement
+                  public void unreachable(ProcessContext c) {
+                    fail("Pipeline should fail to encode a null Long in VarLongCoder");
+                  }
+                }));
+
+    thrown.expectCause(
+        new ThrowableMessageMatcher<IllegalArgumentException>(equalTo("Custom exception "
+            + "element 1")));
+    thrown.expectMessage("Exception occurred while executing transform Step 2");
```
Comment (jkff) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java]: Can you add a test involving a composite transform also?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java [hunk 3]
```diff
@@ -495,8 +495,36 @@ public void unreachable(ProcessContext c) {
...
+  @Test
+  public void testUserException() throws Exception {
+    Pipeline p = getPipeline();
+    PCollection<Long> pcollection = p.apply("Step 1", Create.of(1))
+        .apply("Step 2", ParDo.of(new DoFn<Integer, Long>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            throw new IllegalStateException("Custom exception element " + c.element());
+          }
+        })).setCoder(VarLongCoder.of());
```
Comment (jkff) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java]: `setCoder` here is probably unnecessray.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java (hunk 2 above)
Comment (mairbek) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java]: Added `testUserExceptionCompositeTransform`

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java [hunk 4]
```diff
@@ -495,8 +496,58 @@ public void unreachable(ProcessContext c) {
...
+    p.apply("Step 1", Create.of(1))
+        .apply("Step 2", ParDo.of(new DoFn<Integer, Long>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            throw new IllegalStateException("Custom exception element " + c.element());
+          }
+        }));
+
+    thrown.expectCause(
+        new ThrowableMessageMatcher<IllegalArgumentException>(equalTo("Custom exception "
```
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java]: This generic doesn't actually exist, and even if it did, it wouldn't do anything, right?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ThrowableMessageMatcher.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ */
+package org.apache.beam.runners.direct;
+
+import static org.hamcrest.Matchers.equalTo;
+
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+
+class ThrowableMessageMatcher extends BaseMatcher<Throwable> {
```
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/ThrowableMessageMatcher.java]: This can be a `TypeSafeMatcher<Throwable>`, I believe

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 6]
```diff
@@ -125,9 +126,20 @@
     /**
      * Wraps {@code cause} into a {@link PipelineExecutionException}.
      */
+    public static PipelineExecutionException wrap(String transformName, Throwable cause) {
+      if (Strings.isNullOrEmpty(transformName)) {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: `transformName` should be explicitly `@Nullable` and we should check only vs `null`. Conflating `null` and `""` is strongly discouraged by the Guava authors when they talk about this exact case.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/UserCodeException.java [hunk 7]
```diff
@@ -29,11 +30,18 @@
 public class UserCodeException extends RuntimeException {
 
   public static UserCodeException wrap(Throwable t) {
+    return wrap("", t);
+  }
+
+  public static UserCodeException wrap(String step, Throwable t) {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/UserCodeException.java]: `String step` -> `@Nullable String stepName`, and check only vs `null` instead of `isNullOrEmpty`

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java (hunk 2 above)
Comment (mairbek) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java]: Done

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java (hunk 4 above)
Comment (mairbek) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java]: Removed generic

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ThrowableMessageMatcher.java (hunk 5 above)
Comment (mairbek) [reply runners/direct-java/src/test/java/org/apache/beam/runners/direct/ThrowableMessageMatcher.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 6 above)
Comment (mairbek) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/UserCodeException.java (hunk 7 above)
Comment (mairbek) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/UserCodeException.java]: Done

**Comment by jkff** [comment]
Sorry I forgot about this PR. It looks good, but could you please rebase?

**Comment by jkff** [comment]
I'm going to close this PR for now since Mairbek left Google and the PR was inactive for a while. That said, this is definitely a welcome feature, so if someone else can pick it up from here, it'd be great.
